bpo-31975 (PEP 565): Show DeprecationWarning in __main__

ncoghlan opened the pull request:
- primary change is to add a new default filter entry for
  `default::DeprecationWarning:__main__`
- secondary change is an internal one to cope with plain
  strings in the warning module's internal filter list
  (this avoids the need to create a compiled regex object
  early on during interpreter startup)


<!-- issue-number: bpo-31975 -->
https://bugs.python.org/issue31975
<!-- /issue-number -->

ncoghlan committed:
bpo-31975 (PEP 565): Show DeprecationWarning in __main__

- primary change is to add a new default filter entry for
  'default::DeprecationWarning:__main__'
- secondary change is an internal one to cope with plain
  strings in the warning module's internal filter list
  (this avoids the need to create a compiled regex object
  early on during interpreter startup)

ncoghlan commented on Lib/test/test_warnings/__init__.py:
```diff
@@ -1186,7 +1211,7 @@ def __del__(self):
         # note: "__main__" filename is not correct, it should be the name
         # of the script
-        self.assertEqual(err, b'__main__:7: UserWarning: test')
+        self.assertEqual(err.decode(), '__main__:7: UserWarning: test')
```
This may look unrelated, but comparing these as strings means a proper diff gets printed when the test fails (and it was this test that picked up the problem with the non-regex entry in the internal filter list).

ncoghlan commented:
@brettcannon Something I noticed while making this, is that we don't currently have a test that ensures the C level filter initialisation is consistent with the Python level default filter initialisation at https://github.com/python/cpython/blob/3.6/Lib/warnings.py#L506

I'm thinking this may be a good opportunity to address that, by adding a variant of the test added by the PR that forces the subprocess to reload the Python warnings module without the C accelerator module before printing the contents of the filter list.

ncoghlan commented:
It's also suspicious that `./python -m test test_warnings` passed, even though I hadn't fixed this line yet: https://github.com/python/cpython/blob/3.6/Lib/warnings.py#L344 (similar to the changed C code in the PR, those lines assume that the module name is always either None or a compiled regex object - no strings allowed)

brettcannon commented:
It would probably be easier to refactor that code into a function so that you can write a test which tracks what changed in Python prior to launching.

ncoghlan committed:
Ensure subprocess gets a fresh warnings module

While CPython doesn't implicitly import `warnings` at startup,
some other implementations might do so. Explicitly clearing the
module cache ensures the subprocess also gets the pure Python
version, no matter what.

ncoghlan committed:
PEP 565: Documentation updates

ncoghlan committed:
PEP 565: NEWS entry

ncoghlan committed:
Attempt to make new warnings test work on Windows

ncoghlan committed:
Clarify that this is a list of examples

ncoghlan committed:
Update test_cmd_line for PEP 565

ncoghlan committed:
Fix docs warnings for colon use
